### PR TITLE
Make data sources faster

### LINKED ISSUE #6384: Some contextual actions take too long to finish and there's no progress indication
slow-close 2019-10-29 14_41_13

The app looks unresponsive when you press some actions, namely all actions that don't involve navigation. It's possible to execute other actions while the first action doesn't complete.

What should we do? Simply early closing the menu helps with the look and feel of the view, but the effect of the menu action would take some time to complete e be rendered.

Displaying some sort of progress indicator is also an option.

### PR DESCRIPTION
**Pull request by daividssilverio**
## What's this?
This PR makes two changes:
- It makes the data sources faster
- It fixes a side effect caused by ☝️ in the calendar on droid. iOS still needs to be fixed.

### Relationships
Closes #6384
Supersedes #6765 

## Why do we want this?
Because :sanic:

## How is it done?
All the credit goes to @heytherewill , it basically changes how the data sources report changes in our entities. If I understood it correctly, by making this change, we don't have to wait on everyone who's listening for changes in our entities.

More specifically, these are the interactors that are relying on the `ReportChange` functionality:
```cs
ObserveAllTimeEntriesVisibleToTheUserInteractor.cs
ObserveTimeEntriesChangesInteractor.cs
ObserveTimeEntriesVisibleToTheUserInteractor.cs
ObserveTimeTrackedTodayInteractor.cs
ObserveAllWorkspacesInteractor.cs
ObserveWorkspacesChangesInteractor.cs
```
and of course the `TimeEntriesDataSource` and all related code.

Basically, this affects everything that's listening to changes in Time entries and workspaces... a good deal of the app.

### Why not another way?
By all means, feel free to suggest something different and find flaws in this approach.

### Side effects
There could be a lot. Please help testing it.
From my tests, the calendar view broke on both iOS and android.
I've fixed the android part and I need someone from @toggl/ios-senseis to take a look and fix what I broke.

## Review hints
Review by commit.
Take a look at our `ObservableDataSource` and check the `ReportChange` method.
Check all affected interactors and their usages.

## :squid: Permissions
🦑 after iOS is fixed and @heytherewill ✅ and at least one android and one iOS dev reviews and tests the app and confirms that nothing else is broken.

### CODE REVIEW

**Comment by heytherewill** [comment]
I don't get why you switched to the scheduler provider. This whole view will get 🔥 eventually and things will start relying on database queries, which will be a lot faster. Then, the type that observes changes and reports them back will be lean and can rely on using our injected scheduler provider. 

Until then, using the task to marshall to a background thread is good enough of a change and it's easier to track. Unless that's breaking tests (which it might) I don't see a reason to do everything using the scheduler. Please let me know if tests were broken by the `Task.Run` or not :)

**Comment by daividssilverio** [comment]
@heytherewill 

> This whole view will get 🔥 eventually

 what are you talking about?

> Until then, using the task to marshall to a background thread is good enough of a change and it's easier to track. Unless that's breaking tests (which it might) I don't see a reason to do everything using the scheduler. Please let me know if tests were broken by the Task.Run or not :)

You've guessed it right. Tests become flaky by using using the `Task.Run`, so I've used the scheduler provider.

File: Toggl.Core.UI/ViewModels/Calendar/ContextualMenu/CalendarContextualMenuViewModel.cs [hunk 1]
```diff
@@ -231,7 +231,6 @@ private void closeMenuWithCommittedChanges()
             currentCalendarItem = default;
             calendarItemThatOriginallyTriggeredTheMenu = null;
             currentMenuType = ContextualMenuType.Closed;
-            calendarItemInEditMode.OnNext(null);
```
[the remark this answers is not in the document]
Comment (sofokli1) [reply Toggl.Core.UI/ViewModels/Calendar/ContextualMenu/CalendarContextualMenuViewModel.cs]: If you reintroduce this, this is good for approval.

[the remark this answers is not in the document]
Comment (daividssilverio) [reply Toggl.Core.UI/ViewModels/Calendar/ContextualMenu/CalendarContextualMenuViewModel.cs]: Yeah. Leaving that makes the calendar items being edited blink with old/new values when we save changes.

iOS could do like android and decide to show or not the handles by using the menu visibility 🤔 That should work alright 🤔

**Pull request merged**